mkosi/distributions/opensuse.py

```diff
@@ -21,47 +21,131 @@ def filesystem(cls) -> str:
 
     @classmethod
     def install(cls, state: "MkosiState") -> None:
+        if state.config.base_image:
+            # We assume that the base image has been properly initialized and it
+            # contains all the metadata we need to install the additional
+            # packages.
+            return zypper_install(state, {*state.config.packages})
+
         return install_opensuse(state)
 
+    @classmethod
+    def remove_packages(cls, state: "MkosiState", packages: list[str]) -> None:
+        zypper_remove(state, packages)
+
     @staticmethod
     def initrd_path(kver: str) -> Path:
         return Path("boot") / f"initrd-{kver}"
 
 
+def invoke_zypper(state: MkosiState,
+                  global_opts: set[str],
+                  verb: str,
+                  verb_opts: set[str],
+                  *args: Any,
+                  with_apivfs: bool = False) -> None:
+    zypp_conf = state.root.joinpath("etc/zypp/zypp.conf")
+
+    cmdline = ["zypper", "--root", state.root, *global_opts, verb, *verb_opts, *args]
+
+    env=dict(ZYPP_CONF=zypp_conf)
+
+    if with_apivfs:
+        run_with_apivfs(state, cmdline, env=env)
+    else:
+        run(cmdline, env=env)
+
+
+def zypper_init(state: MkosiState) -> None:
+    state.root.joinpath("etc/zypp").mkdir(mode=0o755, parents=True, exist_ok=True)
+
+    # No matter if --root is used or not, zypper always considers its config
+    # files from the host environment. If we want to use our custom versions for
+    # the rootfs, we're left with two ways to specify them, depending on whether
+    # we need to customize a setting defined in zypp.conf or in zypper.conf. If
+    # it's in zypp.conf then the environment variable 'ZYPP_CONF' must be used
+    # (!) otherwise a custom zypper.conf can be specified with the global option
+    # '--config'.
+
+    zypp_conf = state.root.joinpath("etc/zypp/zypp.conf")
+
+    # For some reason zypper has no command line option to exclude the docs,
+    # this can only be configured via zypp.conf.
+    def yes_no(b: bool) -> str:
+        return "yes" if b else "no"
+
+    zypp_conf_content = f"""\
+    [main]
+    solver.onlyRequires = yes
+    rpm.install.excludedocs = {yes_no(state.config.with_docs)}
+    """
+
+    zypp_conf.write_text(zypp_conf_content)
+
+
+def zypper_addrepo(state: "MkosiState", url: str, name: str, caching: bool = False) -> None:
+    invoke_zypper(state, [], "addrepo", ["-c", "-k" if caching else "-K"], url, name)
+
+
+def zypper_removerepo(state: "MkosiState", repo: str) -> None:
+    invoke_zypper(state, [], "removerepo", [], repo)
+
+
+def zypper_modifyrepo(state: "MkosiState", repo: str, caching: bool = None) -> None:
+    invoke_zypper(state, [], "modifyrepo", ["-k" if caching else "-K"], repo)
+
+
+def zypper_install(state: "MkosiState", packages: set[str]) -> None:
+    global_opts = [
+        "--cache-dir", state.cache,
```

Comment:
```suggestion
        "--cache-dir", os.fspath(state.cache),
```
is the most minimal fix (needs `import os` at the top, too), although all the zypper stuff should probably not take `list[str]`, but `list[PathString]` where appropriate.